lib/components/taskbar/notifications.dart

```diff
@@ -14,18 +14,126 @@ See the License for the specific language governing permissions and
+      child: _NotificationIcon(),
     );
   }
 }
+
+class _NotificationIcon extends StatefulWidget {
+  const _NotificationIcon();
+
+  @override
+  State<_NotificationIcon> createState() => _NotificationIconState();
+}
+
+class _NotificationIconState extends State<_NotificationIcon>
+    with
+        NotificationServiceListener,
+        StateServiceListener<NotificationService, _NotificationIcon> {
+  bool _shellListenerInit = false;
+  bool _unreadNotifs = false;
+
+  ValueNotifier<bool> get notifier =>
+      ShellService.current.getShowingNotifier(NotificationsOverlay.overlayId);
+
+  @override
+  void initState() {
+    super.initState();
+    _unreadNotifs = NotificationService.current.notifications.isNotEmpty;
+  }
+
+  @override
+  void didChangeDependencies() {
+    if (!_shellListenerInit) {
+      _shellListenerInit = true;
+      notifier.addListener(_resetIndicator);
+    }
+    super.didChangeDependencies();
+  }
+
+  @override
+  void dispose() {
+    notifier.removeListener(_resetIndicator);
+    super.dispose();
+  }
+
+  void _resetIndicator() {
+    _unreadNotifs = false;
+    setState(() {});
+  }
+
+  @override
+  Widget buildChild(BuildContext context, NotificationService service) {
+    return Padding(
+      padding: const EdgeInsets.symmetric(horizontal: 4),
+      child: Center(
+        child: ConstrainedBox(
+          constraints: const BoxConstraints(minWidth: 20, maxHeight: 20),
+          child: ValueListenableBuilder<bool>(
+            valueListenable: notifier,
+            builder: (context, showing, _) {
+              return service.notifications.isNotEmpty
+                  ? Material(
+                      color: _unreadNotifs
+                          ? Theme.of(context).colorScheme.secondary
+                          : Theme.of(context).surfaceColor,
+                      type: showing
+                          ? MaterialType.transparency
+                          : MaterialType.canvas,
+                      shape: Constants.circularShape,
+                      child: Center(
+                        child: Padding(
+                          padding: const EdgeInsets.symmetric(horizontal: 4),
+                          child: Text(
+                            service.notifications.length.toString(),
+                            style: TextStyle(
+                              fontSize: 12,
+                              color: _unreadNotifs || showing
+                                  ? Theme.of(context).colorScheme.onSecondary
+                                  : null,
+                            ),
+                          ),
+                        ),
+                      ),
+                    )
+                  : const Icon(
+                      Icons.notifications,
+                      size: 18,
+                    );
+            },
+          ),
+        ),
+      ),
+    );
+  }
+
+  @override
+  void onNotificationAdded(UserNotification id) {
```

Comment:
Consider a different name for the paramter as "id" is often associated with Integers in that class